### PR TITLE
Coordinated

### PR DESCRIPTION
**Pull request by edwardalee**
The coordinated branch now correctly executes simple examples with centralized coordination, including support for physical actions and feedback loops. There are no conflicts with master and all tests pass (on my machine at least), including added tests for the new functionality.

This branch contains substantial enough changes, mostly to the C codebase (federated.c, rti.c, reactor_threaded.c) that it will be almost impossible to do anything in the master branch on these files without creating a conflict. I think that these changes are large enough that a line-by-line code review will take quite a while, and I don't wait that long before merging into master, so I suggest we merge now.

### CODE REVIEW

File: xtext/org.icyphy.linguafranca/src/lib/core/federate.c [hunk 1]
```diff
@@ -54,80 +55,25 @@ char* ERROR_SENDING_MESSAGE = "ERROR sending message to federate via RTI";
...
-
-/**
- * The port used for the server socket 
- * to listen for messages from other federates.
- * The federate informs the RTI of this port once
- * it has created its socket server by sending
- * an ADDRESS_AD message (@see rti.h).
- */
-int _lf_server_port = -1;
+federate_instance_t _fed = {
```
Comment (lhstrh) [inline xtext/org.icyphy.linguafranca/src/lib/core/federate.c:60]: Nice!

File: xtext/org.icyphy.linguafranca/src/lib/core/federate.c [hunk 2]
```diff
@@ -1498,108 +1389,276 @@ void _lf_logical_tag_complete(instant_t time, microstep_t microstep) {
...
-            }
-            // If we get here, any activity on the event queue is not relevant.
-            // Either the queue is empty or whatever appeared on it
-            // has a timestamp greater than this request.
-            // Keep waiting for the TAG.
-        }
+    // What we do next depends on whether the NET has been bounded by
+    // physical time or by an event on the event queue.
+    if (net_bounded_by_physical_time) {
+    	// The NET that was just sent was bounded by physical time because there
```
Comment (lhstrh) [inline xtext/org.icyphy.linguafranca/src/lib/core/federate.c:1540]: Indentation through line 1662 is off.

**Review by lhstrh — APPROVED** [review]
Looks like a ton of clean up and refactoring with some pretty well hidden technical changes. Very nice! I agree we should merge this into master without further ado.

File: xtext/org.icyphy.linguafranca/src/lib/core/federate.c [hunk 3]
```diff
@@ -1498,108 +1389,276 @@ void _lf_logical_tag_complete(instant_t time, microstep_t microstep) {
...
+    // earlier than t. But if there are downstream federates and there is
+    // a physical action (not counting receivers from upstream federates),
+    // then we can only promise up to current physical time (plus the minimum
+    // of all minimum delays on the physical actions).
+    // If wait_for_reply is false, leave the tag alone.
+    bool net_bounded_by_physical_time = wait_for_reply ?
+    		_lf_bounded_NET(&tag)
+			: false;
+
+    send_net(NEXT_EVENT_TIME, tag.time, tag.microstep);
```
Comment (Soroosh129) [inline xtext/org.icyphy.linguafranca/src/lib/core/federate.c:1527]: Should last sent NET be checked here as well?

**Review by Soroosh129 — APPROVED** [review]
The changes look great to me. I agree that they should be merged into master. There are a few FIXMEs in rti.h that I believe can perhaps be addressed after the merge.

**Pull request merged**